### PR TITLE
[RFC] Ray memory improvements: format and summary

### PR DESCRIPTION
**Pull request by micahtyong**
## In this PR

Addresses point 5 in #14428 by modifying `memory_summary` to have more readable formatting when running with `--no-format` or when terminal size is not sufficiently large (e.g. > 137 characters).

Addresses point 6 in #14428 by adding a helper function in `memory_utils.py` that aggregates the amount of memory each reference type uses within a group. 

Addresses point 7 in #14428 by adding support for unit conversions in displaying object sizes (e.g. B, KB, MB, GB) with `ray memory --units=`. 

Addresses point 8 in #14428 by adding the ability to specify the number of entries displayed per group with `ray memory --n`. 

The full view now looks like the following: 
Screen Shot 2021-03-07 at 12 42 52 AM

The unformatted view looks like the following:
Screen Shot 2021-03-07 at 12 42 24 AM

## Related issue number

#14428

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

cc: @rkooo567

### CODE REVIEW

File: dashboard/memory_utils.py [hunk 1]
```diff
@@ -391,10 +427,14 @@ def memory_summary(state,
         mem += f"--- Object references for {group_by}: {key} ---\n"
         mem += object_ref_string\
             .format(*object_ref_labels)
+        n = 0
         for entry in group["entries"]:
+            n += 1
```
Comment (rkooo567) [inline dashboard/memory_utils.py:432]: Can you comment what this line is for?

File: dashboard/memory_utils.py [hunk 2]
```diff
@@ -328,10 +328,22 @@ def construct_memory_table(workers_stats: List,
     return memory_table
 
 
+def track_reference_size(group):
+    """Returns dictionary mapping reference type
+    to memory usage for a given memory table group."""
+    from collections import defaultdict
```
Comment (rkooo567) [inline dashboard/memory_utils.py:334]: Take out this import and move it to the top?

File: dashboard/memory_utils.py [hunk 3]
```diff
@@ -360,27 +375,48 @@ def memory_summary(state,
...
+        Sorting by {sort_by}...\
+        Display {num_entries if num_entries is not None else 'all'}\
+entries per group...\n\n\n"
 
     for key, group in memory_table["group"].items():
         # Group summary
         summary = group["summary"]
-        summary["total_object_size"] = str(summary["total_object_size"]) + " B"
+        ref_size = track_reference_size(group)
+        summary["total_object_size"] = str(
```
Comment (rkooo567) [inline dashboard/memory_utils.py:403]: Can you clean up this part using the for loop?

**Review by rkooo567 — APPROVED** [review]
Nice PR! Just a few minor comments :)!

File: dashboard/memory_utils.py (hunk 1 above)
Comment (micahtyong) [reply dashboard/memory_utils.py:432]: Gotcha. I'll add that `n` is a counter for num entries per group.

**Pull request merged**